src/main/protobuf/soc/playerState.proto

```diff
@@ -0,0 +1,36 @@
+syntax = "proto2";
+
+package protos.soc;
+
+import "soc/inventory.proto";
+
+message PublicInventory {
+    required int32 numCards = 4;
+    required int32 numDevelopmentCards = 5;
+}
+
+message PerfectInventory {
+    required ResourceSet resourceCards = 1;
+    required DevelopmentCardSet playedDevCards = 2;
+    required DevelopmentCardSet canPlayDevCards = 3;
+    required DevelopmentCardSet cannotPlayDevCards = 4;
+}
+
+message ProbableInventory {
+    required ProbableResourceSet probableResourceCards = 1;
+    required DevelopmentCardSet playedDevCards = 2;
+    //TODO Add ProbableDevelopment cards
+}
+
+message PlayerState {
```

Comment:
So here is my take on how we want to represent state, despite what I've said earlier, let's forget about probable state right now. It really confuses the issue, and there is a way to weave it in in the future, if necessary, but really let's forget about it for now. 

A) Clients can track it themselves
B) A client can always make a request to the server `GetProbableInvetory(gameId, perspectivePlayerPostion)
ie. `getProbablyInventory(gamId = 10, perspectivPlayerPostion = 2) would return a map of the probable inventory of the other 3 players from player 2's perspective.

Reply:
Okay, now that we can safely ignore probable inventory, let's talk about the important part. `GameState`. 

It gets passed along with every action request. Right now, while action requests are only sent to individual players, I still like to envision the `GameState` as being a public object that all players can see. Yes, this means encryption.

I did a bit of research, and mutli encryption does not seem so easy (though still have to more research), instead I propose the following `HiddenCard` defintion:

```
message HiddenCard {
   repeated int32 viewable_by_positions = 1
   required publicEncryptedValue = 2 // the publicEncyrptedValue is associated with a symmetric game key, so a "SuperObserver" or researcher can easily decode game state at any turn. 
   required map<int32, string> encryptedValues = 3 // map from each players position to the encrypted value. 
}

Reply:
Now that we have a `HiddenCard` defined, let's go about defining the `GameState`... the only thing that really changes from this PR, is the `PlayerState` object. Here is how I would define it:

```
message PlayerState {
    required string name = 1;
    required int32 position = 2;
    required int32 numPoints = 3;
    required int32 numRoads = 4;
    required int32 numKnights = 5;
    required bool hasLongest = 6;
    required bool hasLargest = 7;
    repeated DevelopmentCard playedDevelopmentCards = 8;
    repeated map<int, HiddenCard> developmentCards = 9; // map from turn purchased to unplayed development card
    repeated HiddenCard resourceInventory = 9;
}

Reply:
And that's it, yes clients have to decode their hidden cards, but there is such beautiful parallelism. The `GameState` can be passed between players safely, it's a one size fits all object. And I think it's more intuitive than baking in all the logic of what's visible and what's not. It's like a real game. You see everyone's cards, you just can "decrypt" them. The state is represented the same way.

Reply:
I really do not like the idea of encryption. I am fine with a public `GameState` but I do not see why that would need to include private information. I still support that with each request only sending the request to the individual player and that can include their private inventory and the public gamestate. The private inventory can be seperate from the public gamestate

Reply:
You would still see if a player bought a development card. You would just have no way of seeing what it is

Reply:
And all the logic of what is visible and what is already baked in. There is no new logic to write

Reply:
It's not about what's already baked in, it's about what's going to be an easier data model to work with. I think an encryption pattern allows us to do that, but I also think it's okay to send separate messages, and that pattern will work too. Let's talk about it tonight, but ultimately I'll defer to you.

Reply:
Ahh, honestly I don't know. Encryption is a mess. Let's talk tonight about pros and cons.

Reply:
Though also just to pull back the curtain, it's actually not a lot of work for a client: 

Let's say I want to decrypt a hidden resource:
`resource = Resource[decrypt(payload.hiddenCard[position], myKey)]` It's oneline, super easy.

Reply:
but there is a lot of overhead of sharing public keys and having to remake public keys which all has to be done by client and server. Its that part that would be the mess

Reply:
In light of our conversation what are your thoughts on this for `PlayerState`:

```
message HiddenCard {
   repeated int32 view_by_positions = 1
   oneof {
      Resource resource = 2;
      DevelopmentCard developmentCard = 3;
   }
   int32 optional turnPurchased = 4;
}

message PlayerState {
    required string name = 1;
    required int32 position = 2;
    required int32 boardPoints = 3;
    required int32 hiddenPoints = 4; // this would always be 0 unless the state is passed to the given player
    required bool hasLongest = 5;
    required bool hasLargest = 6;
    repeated map<int, DevelopmentCard> playedDevelopmentCards = 8; // map from turn played to the card
    repeated HiddenCard developmentCards = 9; 
    repeated HiddenCard resourceInventory = 9;
}
```

Note, I removed the longest road and army. For army, very easy for client to count:
`knightCount = len([card == KNIGHT for card in playedDevelopmentCards.values])`

For road, it's more complicated, but that why I suggest, like for probable hands, we create an RPC, like `getLongestRoadInformation`

Which returns a whole obejct that shows each players road count, the path to their possible longest, and other metadata if we choose.

Reply:
We have to include flags for hasLargest or HasLongest because two players can have the same amount of knights or roads and but one has longest and one does not

Reply:
Good point, the `PlayerState` above includes `hasLargest` and `hasLongest`, it just doesn't have the actual counts. What is your thought for my proposed `PlayerState` above?